File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.py
import sys
input = sys.stdin.readline

while True:
    stack=[]
    str = input().rstrip()
    if str=='.':
        break
    for i in str:
        if i=='(' or i=='[':
            stack.append(i)
        elif i==')':
            if not stack or stack[-1]!='(':
                stack.append(i)
                break
            else:
                stack.pop()
        elif i==']':
            if not stack or stack[-1]!='[':
                stack.append(i)
                break
            else:
                stack.pop()
    if len(stack) == 0:
        print('yes')
    else:
        print('no')

File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Stack;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();
        Stack<Character> stack;
        String endFlag = ".";

        while(true){
            String str = br.readLine();
            if(str.equals(endFlag)) {
                System.out.println(sb);
                return;
            }
            stack = new Stack<>();

            for(int i=0; i<str.length(); i++){
                if(str.charAt(i) == '(' || str.charAt(i)=='[') {
                    stack.add(str.charAt(i));
                } else if(str.charAt(i)==')'){
                    if(stack.isEmpty() || stack.pop()!='('){
                        stack.add(str.charAt(i));
                        break;
                    }
                }
                else if(str.charAt(i)==']'){
                    if(stack.isEmpty() || stack.pop()!='['){
                        stack.add(str.charAt(i));
                        break;
                    }
                }
            }
            if(stack.isEmpty()){
                sb.append("yes").append("\n");
            } else {
                sb.append("no").append("\n");
            }
        }
    }
}
